stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/attachments/AttachmentController.kt

```diff
@@ -0,0 +1,110 @@
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+internal class AttachmentController(
+    private val context: Context,
+    private val selectedMediaRecyclerView: RecyclerView,
+    private val selectedFilesRecyclerView: RecyclerView,
+    private val selectedAttachmentsChangeListener: (List<AttachmentMetaData>) -> Unit
+
+) {
+    internal var selectedAttachments: List<AttachmentMetaData> = emptyList()
+        private set
+
+    private val storageHelper = StorageHelper()
+    private val selectedFileAttachmentAdapter = SelectedFileAttachmentAdapter(::cancelAttachment)
+    private val selectedMediaAttachmentAdapter = SelectedMediaAttachmentAdapter(::cancelAttachment)
+
+    private val attachmentSelectionListener = object : AttachmentSelectionListener {
+        override fun onAttachmentsSelected(attachments: Set<AttachmentMetaData>, attachmentSource: AttachmentSource) {
+            if (attachments.isNotEmpty()) {
+                when (attachmentSource) {
+                    AttachmentSource.MEDIA,
+                    AttachmentSource.CAMERA -> onMediaAttachmentsSelected(attachments)
+                    AttachmentSource.FILE -> onFileAttachmentsSelected(attachments)
+                }
+            }
+        }
+    }
+
+    init {
+        selectedFilesRecyclerView.itemAnimator = null
+        selectedFilesRecyclerView.adapter = selectedFileAttachmentAdapter
+        selectedMediaRecyclerView.adapter = selectedMediaAttachmentAdapter
+    }
+
+    fun openAttachmentDialog() {
+        context.getFragmentManager()?.let {
+            AttachmentDialogFragment.newInstance()
+                .apply { setAttachmentSelectionListener(attachmentSelectionListener) }
+                .show(it, AttachmentDialogFragment.TAG)
+        }
+    }
+
+    fun clearSelectedAttachments() {
+        selectedAttachments = emptyList()
+        selectedFilesRecyclerView.isVisible = false
+        selectedFileAttachmentAdapter.clear()
+        selectedMediaRecyclerView.isVisible = false
+        selectedMediaAttachmentAdapter.clear()
+    }
+
+    private fun cancelAttachment(attachment: AttachmentMetaData) {
+        selectedAttachments = selectedAttachments - attachment
+        selectedFileAttachmentAdapter.removeAttachment(attachment)
+        selectedMediaAttachmentAdapter.removeAttachment(attachment)
+
+        if (selectedAttachments.isEmpty()) {
+            selectedFilesRecyclerView.isVisible = false
+            selectedFileAttachmentAdapter.clear()
+            selectedMediaRecyclerView.isVisible = false
+            selectedMediaAttachmentAdapter.clear()
+        }
+
+        selectedAttachmentsChangeListener(selectedAttachments)
+    }
+
+    private fun onMediaAttachmentsSelected(attachments: Set<AttachmentMetaData>) {
+        selectedAttachments = attachments
+            .filter { it.uri != null }
+            .filter { it.type in listOf(ModelType.attach_image, ModelType.attach_video) }
```

Comment:
Nit: Maybe declare in companion such list of applicable attachments type?
``` kotlin
selectedAttachments = attachments
            .filter { it.uri != null }
            .filter { it.type in MEDIA_ATTACHMENTS_TYPES }
...
companion object {
  private val MEDIA_ATTACHMENTS_TYPES = listOf(ModelType.attach_image, ModelType.attach_video)
}
```

I think it could be more efficient in memory consumption with big amount of data

Reply:
Really good catch

Reply:
Done